internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,511 @@
+func (m *TLSManager) TLSKeyPair() (*TLSKeyPair, error) {
+	m.tlsKeyPairLock.Lock()
+	defer m.tlsKeyPairLock.Unlock()
+
+	var selectedKeypair *TLSKeyPair
+	// Try to find an existing valid certificate.
+	for i, keypair := range m.tlsKeyPairs {
+		if !keypair.isExpired() && m.isTLSKeyPairSupported(keypair) {
+			selectedKeypair = m.tlsKeyPairs[i]
+			break
+		}
+	}
+	// Generate a keypair if no one exists.
+	if selectedKeypair == nil {
+		var err error
+		notBefore := systemTime()
+		notAfter := notBefore.Add(certificateValidityPeriod)
+		selectedKeypair, err = m.addNewTLSKeyPair(notBefore, notAfter)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Return the TLS keypair.
+	return selectedKeypair, nil
+}
+
+// TLSKeyPairs returns the slice of currently valid and supported tlsKeyPairs.
+func (m *TLSManager) TLSKeyPairs() []*TLSKeyPair {
+	m.tlsKeyPairLock.Lock()
+	defer m.tlsKeyPairLock.Unlock()
+
+	var activeTLSKeyPairs []*TLSKeyPair
+	for _, keypair := range m.tlsKeyPairs {
+		if !keypair.isExpired() && m.isTLSKeyPairSupported(keypair) {
+			activeTLSKeyPairs = append(activeTLSKeyPairs, keypair)
+		}
+	}
+	return activeTLSKeyPairs
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *TLSManager) Ensure() error {
+	return nil
+}
+
+// loadTLSKeypairs loads all the keypairs from the TLS directory. This
+// method can be called again to reload the directory during run-time
+// as long as the tlsKeyPairLock is locked is held. The keypairs are ordered
+// by the certificate start date, then end date.
+func (m *TLSManager) loadTLSKeyPairs() error {
+	var keypairs []*TLSKeyPair
+	err := m.withPEMFile(func(path string, fingerprint string) error {
+		keypair := &TLSKeyPair{}
+		err := keypair.fromPEMFile(path)
+		if err != nil {
+			return fmt.Errorf("cannot load keypair from PEM file %q: %w", path, err)
+		}
+		keypair.addFingerprint()
+		if keypair.fingerprint != fingerprint {
+			return fmt.Errorf("cannot match fingerprint with filename: %q vs. %q", keypair.fingerprint, fingerprint)
+		}
+
+		keypairs = append(keypairs, keypair)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	// Sort by valid starting date, end date and fingerprint.
+	slices.SortFunc(keypairs, func(a, b *TLSKeyPair) int {
+		return a.Compare(b)
+	})
+	m.tlsKeyPairs = keypairs
+	return nil
+}
+
+// filenameRegexp matches PEM files based on an SHA512/384 hash of the
+// x509 raw certificate, encoded in base32 (77 digits), with an ".pem" suffix.
+var filenameRegexp = regexp.MustCompile(`^[ABCDEFGHIJKLMNOPQRSTUVWXYZ234567]{77}\.pem$`)
+
+// withPEMFile inspects the TLS directory, locates all the PEM files and
+// call the 'do' function on each valid file. A non-existant or empty TLS
+// directory is valid, but a TLS directory or PEM file with invalid
+// permissions will result in an error. Files in the TLS directory not
+// matching the naming convention will be ignored.
+func (m *TLSManager) withPEMFile(do func(path string, fingerprint string) error) error {
```

Comment:
I don't think the callback is necessary, especially considering this function is only used in one context.

You could also have one function that returns a slice of all the fingerprints found in the certificate directory. Then in a different function you can iterate on that slice, opening the files one by one.

Reply:
Removed. Resolving.